Add GetPinSecret to protocol

patrickxb opened the pull request:
Added a new func to SecretUI protocol for integration w/ electron pinentry.

The new argument is:

    record PinEntryArg {
        string windowTitle;
        string prompt;
        boolean allowSecretStorage;
        string secretStorageLabel;
    }

and it returns GetPassphraseRes.

Nothing calls this yet, but if the desktop app registers as a secret ui delegate and a change is made to call GetPinSecret() instead of GetSecret() somewhere, integration with the electron pinentry should be testable.

r? @cjb

cjb commented:
@patrickxb LGTM!

Would it be worth adding a temporary, separate command that invokes GetPinSecret() for testing?

maxtaco commented:
Can we call it something other than `GetPinSecret`?

maxtaco commented:
I.e., the whole idea of calling it "pinentry" is pretty crazy.  That's like 0123.  Since we're replacing it though, can we do away with the outmoded terminology?

patrickxb commented:
@cjb Sure.  Will do.

Also, a lot of this cruft can get cleaned out once this is working.......

patrickxb commented:
@maxtaco would be happy to.  ideas?

maxtaco commented:
Hmm, we already have ~4 things named very similarly.

cjb commented:
@patrickxb The JSON blob I linked to had a 'features' array, which would let you have multiple checkboxes.  I don't think we need multiple checkboxes today, but it might still be worth getting the plumbing there.  (My JS-side handles it already.)  What do you think?

patrickxb commented:
@maxtaco once this is working, I think we can consolidate.

cjb commented:
@patrickxb I don't think we can remove pinentry after this is added, though? Because some users won't have an Electron delegate connected for whatever reason.

maxtaco commented:
In which case, maybe `getPassphrase` or if that clashes with an old RPC, then `getPasssphrase2`

patrickxb commented:
@cjb if they don't it will use the terminal

patrickxb commented:
@maxtaco: I changed it to getPassphrase.
@cjb: I added a features record, test-passphrase command.  Tested it and it did work on terminal...

So I think this is ready to go.  PTAL.

maxtaco commented on go/systests/secret_ui_test.go:
```diff
@@ -88,6 +88,7 @@ type secretUI struct {
 	getNewPassphrase      bool
 	getPaperKeyPassphrase bool
 	getSecret             bool
+	getPinSecret          bool
```
rename?

maxtaco commented on go/systests/secret_ui_test.go:
```diff
@@ -117,6 +118,11 @@ func (s *secretUI) GetSecret(context.Context, keybase1.GetSecretArg) (res keybas
 }
 
+func (s *secretUI) GetPassphrase(context.Context, keybase1.GetPassphraseArg) (res keybase1.GetPassphraseRes, err error) {
+	s.getPinSecret = true
```
rename?

maxtaco commented on go/systests/user_test.go:
```diff
@@ -91,6 +91,11 @@ func (n *signupSecretUI) GetSecret(pinentry keybase1.SecretEntryArg, terminal *k
 }
 
+func (n *signupSecretUI) GetPassphrase(p keybase1.GUIEntryArg, terminal *keybase1.SecretEntryArg) (res keybase1.GetPassphraseRes, err error) {
+	err = fmt.Errorf("GetPinSecret unimplemented")
```
use new name?

maxtaco replied:
change name pls?

maxtaco commented on protocol/avdl/secret_ui.avdl:
```diff
@@ -36,4 +36,22 @@ protocol secretUi {
+
+  record GUIEntryArg {
+    string windowTitle;
+    string prompt;
```
we'll also need the equivalent of `retryLabel` too, won't we?

patrickxb replied:
this is based on @cjb doc: https://github.com/keybase/client/blob/8d718a207f1d4e64f1bb6ddfac2325adf0f7da0a/desktop/docs/electron_architecture.md

can just change the prompt during retry?

maxtaco replied:
Well you might lose the ability to style in that case.

cjb replied:
Hrr, adding `retryLabel` probably is cleaner, so that we can call attention to it.  Sorry for not getting this right the first time, @patrickxb!

patrickxb replied:
Ok, it's in

maxtaco commented on go/libkb/test_common.go:
```diff
@@ -337,6 +337,7 @@ type TestSecretUI struct {
 	CalledGetKBPassphrase  bool
 	CalledGetBUPassphrase  bool
 	CalledGetNewPassphrase bool
+	CalledGetPinSecret     bool
```
rename/

maxtaco commented on go/libkb/test_common.go:
```diff
@@ -363,6 +364,14 @@ func (t *TestSecretUI) GetPaperKeyPassphrase(keybase1.GetPaperKeyPassphraseArg)
 }
 
+func (t *TestSecretUI) GetPassphrase(p keybase1.GUIEntryArg, terminal *keybase1.SecretEntryArg) (keybase1.GetPassphraseRes, error) {
+	t.CalledGetPinSecret = true
```
ditto

maxtaco commented:
:+1:
